core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -453,58 +486,64 @@ public Object visitGetKeysInGroupCommand(final InvocationContext ctx, GetKeysInG
 
    @Override
    public Object visitReadOnlyKeyCommand(InvocationContext ctx, ReadOnlyKeyCommand command) throws Throwable {
+      CompletionStage<Void> stage;
       if (command instanceof TxReadOnlyKeyCommand) {
          // TxReadOnlyKeyCommand may apply some mutations on the entry in context so we need to always wrap it
-         entryFactory.wrapEntryForWriting(ctx, command.getKey(), command.getSegment(), ignoreOwnership(command) || canRead(command), true);
+         stage = entryFactory.wrapEntryForWriting(ctx, command.getKey(), command.getSegment(), ignoreOwnership(command) || canRead(command), true);
       } else {
-         entryFactory.wrapEntryForReading(ctx, command.getKey(), command.getSegment(), ignoreOwnership(command) || canRead(command));
+         stage = entryFactory.wrapEntryForReading(ctx, command.getKey(), command.getSegment(), ignoreOwnership(command) || canRead(command));
       }
 
-      // Repeatable reads are not achievable with functional commands, as we don't store the value locally
-      // and we don't "fix" it on the remote node; therefore, the value will be able to change and identity read
-      // could return different values in the same transaction.
-      // (Note: at this point TX mode is not implemented for functional commands anyway).
-      return invokeNext(ctx, command);
+      if (CompletionStages.isCompletedSuccessfully(stage)) {
```

Comment:
`asyncInvokeNext` already checks if the delay stage is completed.

Reply:
Ah, yes it does :)